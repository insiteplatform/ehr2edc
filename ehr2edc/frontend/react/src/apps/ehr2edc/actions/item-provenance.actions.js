import {ActionType} from 'redux-promise-middleware';

import {getItemProvenance, getSubmittedItemProvenance} from "../../../api/api-ext";

export const FETCH_ITEM_PROVENANCE = 'FETCH_ITEM_PROVENANCE';
export const FETCH_ITEM_PROVENANCE_PENDING = `${FETCH_ITEM_PROVENANCE}_${ActionType.Pending}`;
export const FETCH_ITEM_PROVENANCE_FULFILLED = `${FETCH_ITEM_PROVENANCE}_${ActionType.Fulfilled}`;
export const FETCH_ITEM_PROVENANCE_REJECTED = `${FETCH_ITEM_PROVENANCE}_${ActionType.Rejected}`;
export const FETCH_SUBMITTED_ITEM_PROVENANCE = 'FETCH_SUBMITTED_ITEM_PROVENANCE';
export const FETCH_SUBMITTED_ITEM_PROVENANCE_PENDING = `${FETCH_SUBMITTED_ITEM_PROVENANCE}_${ActionType.Pending}`;
export const FETCH_SUBMITTED_ITEM_PROVENANCE_FULFILLED = `${FETCH_SUBMITTED_ITEM_PROVENANCE}_${ActionType.Fulfilled}`;
export const FETCH_SUBMITTED_ITEM_PROVENANCE_REJECTED = `${FETCH_SUBMITTED_ITEM_PROVENANCE}_${ActionType.Rejected}`;
export const OPEN_ITEM_DETAILS = 'OPEN_ITEM_DETAILS';
export const CLOSE_ITEM_DETAILS = 'CLOSE_ITEM_DETAILS';

export function fetchItemProvenance(payload = {}) {
    return {
        type: FETCH_ITEM_PROVENANCE,
        payload: getItemProvenance(payload.studyId, payload.subjectId, payload.eventId, payload.itemId)
    }
}

export function fetchSubmittedItemProvenance(payload = {}) {
    return {
        type: FETCH_SUBMITTED_ITEM_PROVENANCE,
        payload: getSubmittedItemProvenance(payload.studyId, payload.subjectId, payload.eventId, payload.itemId)
    }
}

export function openItemDetails(payload = {}) {
    return {
        type: OPEN_ITEM_DETAILS,
        payload
    }
}

export function closeItemDetails() {
    return {
        type: CLOSE_ITEM_DETAILS
    }
}