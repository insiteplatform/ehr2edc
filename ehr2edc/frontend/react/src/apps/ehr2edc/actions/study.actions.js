import {ActionType} from 'redux-promise-middleware';
import {getAll} from "../../../api/study-api";

export const FETCH_STUDIES = 'FETCH_STUDIES';
export const FETCH_STUDIES_PENDING = `${FETCH_STUDIES}_${ActionType.Pending}`;
export const FETCH_STUDIES_FULFILLED = `${FETCH_STUDIES}_${ActionType.Fulfilled}`;
export const FETCH_STUDIES_REJECTED = `${FETCH_STUDIES}_${ActionType.Rejected}`;

export function fetchStudies() {
    return {
        type: FETCH_STUDIES,
        payload: getAll()
    };
}