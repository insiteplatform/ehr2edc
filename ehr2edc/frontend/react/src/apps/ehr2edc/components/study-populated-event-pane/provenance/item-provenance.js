import React from "react";
import RequestError from "../../../../../common/components/error/backend-request-error";

export class ItemProvenance extends React.Component {

    render() {
        const {item} = this.props;
        return !!item ? <>
            <h4 className="facts-title">Source Data Details</h4>
            {this.info(item)}
            {this.provenance(item)}
        </> : null;
    }

    info({name, value}) {
        return <dl className="margin-none">
            <h5>Name</h5>
            <dd>{name}</dd>
            <h5>Submitted Value</h5>
            <dd>{value}</dd>
        </dl>;
    }

    provenance({provenance}) {
        if (!!provenance) {
            const {error, data} = provenance;
            return <>
                <h5>Source Data</h5>
                {!error && !data ? this.provenanceLoading() : this.provenanceLoaded(provenance)}
            </>;
        }
    }

    provenanceLoading() {
        return <p>Loading...</p>
    }

    provenanceLoaded(provenance) {
        const {error, data} = provenance;
        return <>
            {this.provenanceError(error)}
            {this.provenanceDataListing(data)}
        </>;
    }

    provenanceError(error) {
        return !!error ?
            <div className="alert-box alert warning margin-trl">
                <RequestError error={error}/>
            </div> : null;
    }

    provenanceDataListing(data) {
        return !!data ? <div>
            {this.renderGroups(data.groups)}
            {this.renderItems(data.items)}
        </div> : null;
    }

    renderGroups(groups) {
        return groups.map((group, idx) => {
            const {label, items} = group;
            return <div key={idx} className="row">
                <h6>{label}</h6>
                {this.renderItems(items)}
            </div>
        })
    }

    renderItems(items) {
        return <dl className="responsive">
            {items.map((item, idx) => {
                const {label, value} = item;
                return <div key={idx} className="row">
                    <dt>{label}</dt>
                    <dd>{this.renderItemValue(value)}</dd>
                </div>;
            })}
        </dl>
    }

    renderItemValue(value) {
        return !!value ? value : <i className="disabled">none</i>
    }
}