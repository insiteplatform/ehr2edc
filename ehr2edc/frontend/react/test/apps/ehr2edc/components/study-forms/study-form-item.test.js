import React from 'react';
import {
    Item,
    mapDispatchToProps,
    mapStateToProps
} from "../../../../../src/apps/ehr2edc/components/study-forms/study-form-item"
import {closeItemDetails, openItemDetails} from "../../../../../src/apps/ehr2edc/actions/item-provenance.actions";

const item = {
    itemKey: "item-key-1",
    id: "id-1",
    subjectId: "subject-1",
    eventId: "event-1",
    name: "Item name",
    value: "Item value",
};
const itemNotSubmittedToEdc = Object.assign({}, item, {submittedToEdc: false});
const itemInReview = Object.assign({}, item, {inReview: true});
const itemReviewed = Object.assign({}, item, {inReview: false});

describe("An Item in review gets rendered well", () => {
    test("A regular Item in review gets rendered well", () => {
        const app = shallow(<Item {...itemInReview}/>);

        expect(app).toMatchSnapshot();
    });

    test("A read-only Item in review gets rendered well", () => {
        const app = shallow(<Item {...itemInReview}
                                  readOnly={true}/>);

        expect(app).toMatchSnapshot();
    });

    test("An Item in review with a unit gets rendered well", () => {
        const app = shallow(<Item {...itemInReview}
                                  unit="kg"/>);

        expect(app).toMatchSnapshot();
    });

    test("An Item in review with a labeled value gets rendered well", () => {
        const app = shallow(<Item {...itemInReview}
                                  valueLabel="Item value label"/>);

        expect(app).toMatchSnapshot();
    });

    test("A selected Item in review gets rendered well", () => {
        const app = shallow(<Item {...itemInReview}
                                  selected={true}/>);

        expect(app).toMatchSnapshot();
    });

    test("A Item with open details is rendered as active", () => {
        const app = shallow(<Item {...itemInReview}
                                  detailsOpened={true}/>);

        expect(app.hasClass("is-active")).toBe(true);
    });

    test("An reviewed Item that was not submitted to edc", () => {
        const app = shallow(<Item {...itemNotSubmittedToEdc}
                                  valueLabel="Item value label"/>);

        expect(app).toMatchSnapshot();
    });
});

describe("A reviewed Item gets rendered well", () => {
    test("A regular reviewed Item gets rendered well", () => {
        const app = shallow(<Item {...itemReviewed}/>);

        expect(app).toMatchSnapshot();
    });

    test("A read-only reviewed Item gets rendered well", () => {
        const app = shallow(<Item {...itemReviewed}
                                  readOnly={true}/>);

        expect(app).toMatchSnapshot();
    });

    test("An reviewed Item with a unit gets rendered well", () => {
        const app = shallow(<Item {...itemReviewed}
                                  unit="kg"/>);

        expect(app).toMatchSnapshot();
    });

    test("A reviewed Item with open details is rendered as active", () => {
        const app = shallow(<Item {...itemReviewed}
                                  detailsOpened={true}/>);

        expect(app.hasClass("is-active")).toBe(true);
    });
});

describe("Actions are dispatched correctly", () => {
    test("The openItemDetails action is dispatched when clicking the button", () => {
        const preventDefault = jest.fn();
        const event = {preventDefault: preventDefault};
        const openItemDetails = jest.fn();
        const actions = {openItemDetails: openItemDetails};
        const app = shallow(<Item {...itemInReview}
                                  actions={actions}/>);

        app.find('.open-item-details').simulate('click', event);

        const expectedPayload = {
            "eventId": itemInReview.eventId,
            "itemId": itemInReview.id,
            "subjectId": itemInReview.subjectId,
            "inReview": itemInReview.inReview,
            "name": itemInReview.name,
            "value": itemInReview.value
        };
        expect(preventDefault).toHaveBeenCalled();
        assertAction(openItemDetails, expectedPayload);
    });

    test("The closeItemDetails action is dispatched when clicking the button and the details for this item is already open", () => {
        const preventDefault = jest.fn();
        const event = {preventDefault: preventDefault};
        const closeItemDetails = jest.fn();
        const actions = {closeItemDetails: closeItemDetails};
        const app = shallow(<Item {...itemInReview}
                                  actions={actions}
                                  detailsOpened={true}/>);

        app.find('.open-item-details').simulate('click', event);

        expect(preventDefault).toHaveBeenCalled();
        assertAction(closeItemDetails);
    });

    test("The onItemSelected action is dispatched correctly", () => {
        const action = jest.fn();
        const app = shallow(<Item {...itemInReview}
                                  onItemSelected={action}/>);

        app.find(`#select-form-item-${item.itemKey}`).simulate("change", {target: {checked: true}});

        assertAction(action, item.itemKey);
    });

    test("The onItemDeselected action is dispatched correctly", () => {
        const action = jest.fn();
        const app = shallow(<Item {...itemInReview}
                                  selected={true}
                                  onItemDeselected={action}/>);

        app.find(`#select-form-item-${item.itemKey}`).simulate("change", {target: {checked: false}});

        assertAction(action, item.itemKey);
    });

    function assertAction(action, expectedPayload) {
        expect(action).toHaveBeenCalledTimes(1);
        if (expectedPayload) {
            expect(action).toHaveBeenCalledWith(expectedPayload);
        }
    }
});

describe("mapDispatchToProps", () => {
    test("openItemDetails", () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).actions.openItemDetails();

        expect(dispatch).toBeCalledWith(openItemDetails());
    });

    test("closeItemDetails", () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).actions.closeItemDetails();

        expect(dispatch).toBeCalledWith(closeItemDetails());
    });
});

describe("mapStateToProps", () => {
    test("Empty state", () => {
        const ownProps = item;
        const state = {};

        expect(mapStateToProps(state, ownProps)).toEqual({detailsOpened: undefined});
    });
    test("No itemprovenance", () => {
        const ownProps = item;
        const state = {itemProvenance: undefined};

        expect(mapStateToProps(state, ownProps)).toEqual({detailsOpened: undefined});
    });
    test("Empty itemprovenance", () => {
        const ownProps = item;
        const state = {itemProvenance: {}};

        expect(mapStateToProps(state, ownProps)).toEqual({detailsOpened: undefined});
    });
    test("No selectedItem", () => {
        const ownProps = item;
        const state = {
            itemProvenance: {
                itemId: undefined
            }
        };

        expect(mapStateToProps(state, ownProps)).toEqual({detailsOpened: undefined});
    });
    test("Different selectedItem", () => {
        const ownProps = item;
        const state = {
            itemProvenance: {
                itemId: `other-${item.id}`
            }
        };

        expect(mapStateToProps(state, ownProps)).toEqual({detailsOpened: false});
    });
    test("Is selectedItem", () => {
        const ownProps = item;
        const state = {
            itemProvenance: {
                itemId: item.id
            }
        };

        expect(mapStateToProps(state, ownProps)).toEqual({detailsOpened: true});
    });
});