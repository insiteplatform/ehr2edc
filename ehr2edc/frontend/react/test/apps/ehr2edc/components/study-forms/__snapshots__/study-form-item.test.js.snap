// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A reviewed Item gets rendered well A read-only reviewed Item gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <div
    className="disabled padding-small-rl"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <small
      style={
        Object {
          "display": "flex",
          "width": "100%",
        }
      }
    >
      <ErrorTooltip
        additionalClassNames={
          Array [
            "small-8",
          ]
        }
      >
        <span
          className="small-8"
        >
          Item name
        </span>
      </ErrorTooltip>
      <span
        className="small-4 text-right"
      >
        Item value
        <a
          className="open-item-details margin-small-left"
          onClick={[Function]}
          title="Open item details"
        >
          <i
            className="fa fa-file-text-o"
          />
        </a>
      </span>
    </small>
  </div>
</div>
`;

exports[`A reviewed Item gets rendered well A regular reviewed Item gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <div
    className="disabled padding-small-rl"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <small
      style={
        Object {
          "display": "flex",
          "width": "100%",
        }
      }
    >
      <ErrorTooltip
        additionalClassNames={
          Array [
            "small-8",
          ]
        }
      >
        <span
          className="small-8"
        >
          Item name
        </span>
      </ErrorTooltip>
      <span
        className="small-4 text-right"
      >
        Item value
        <a
          className="open-item-details margin-small-left"
          onClick={[Function]}
          title="Open item details"
        >
          <i
            className="fa fa-file-text-o"
          />
        </a>
      </span>
    </small>
  </div>
</div>
`;

exports[`A reviewed Item gets rendered well An reviewed Item with a unit gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <div
    className="disabled padding-small-rl"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <small
      style={
        Object {
          "display": "flex",
          "width": "100%",
        }
      }
    >
      <ErrorTooltip
        additionalClassNames={
          Array [
            "small-8",
          ]
        }
      >
        <span
          className="small-8"
        >
          Item name
          <small>
             (
            kg
            )
          </small>
        </span>
      </ErrorTooltip>
      <span
        className="small-4 text-right"
      >
        Item value
        <a
          className="open-item-details margin-small-left"
          onClick={[Function]}
          title="Open item details"
        >
          <i
            className="fa fa-file-text-o"
          />
        </a>
      </span>
    </small>
  </div>
</div>
`;

exports[`An Item in review gets rendered well A read-only Item in review gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
  title="This data point was populated to facilitate review only and cannot be exported (it will not be sent to any EDC system linked to this study)"
>
  <div
    className="disabled padding-small-rl"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <small
      style={
        Object {
          "display": "flex",
          "width": "100%",
        }
      }
    >
      <ErrorTooltip
        additionalClassNames={
          Array [
            "small-8",
          ]
        }
      >
        <span
          className="small-8"
        >
          Item name
        </span>
      </ErrorTooltip>
      <span
        className="small-4 text-right"
      >
        Item value
        <a
          className="open-item-details margin-small-left"
          onClick={[Function]}
          title="Open item details"
        >
          <i
            className="fa fa-file-text-o"
          />
        </a>
      </span>
    </small>
  </div>
</div>
`;

exports[`An Item in review gets rendered well A regular Item in review gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <input
    id="select-form-item-item-key-1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="inline-label"
    htmlFor="select-form-item-item-key-1"
  >
    <ErrorTooltip
      additionalClassNames={
        Array [
          "small-8",
        ]
      }
    >
      <span
        className="small-8"
      >
        Item name
      </span>
    </ErrorTooltip>
    <span
      className="small-4 text-right"
    >
      Item value
      <a
        className="open-item-details margin-small-left"
        onClick={[Function]}
        title="Open item details"
      >
        <i
          className="fa fa-file-text-o"
        />
      </a>
    </span>
  </label>
</div>
`;

exports[`An Item in review gets rendered well A selected Item in review gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <input
    checked={true}
    id="select-form-item-item-key-1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="inline-label"
    htmlFor="select-form-item-item-key-1"
  >
    <ErrorTooltip
      additionalClassNames={
        Array [
          "small-8",
        ]
      }
    >
      <span
        className="small-8"
      >
        Item name
      </span>
    </ErrorTooltip>
    <span
      className="small-4 text-right"
    >
      Item value
      <a
        className="open-item-details margin-small-left"
        onClick={[Function]}
        title="Open item details"
      >
        <i
          className="fa fa-file-text-o"
        />
      </a>
    </span>
  </label>
</div>
`;

exports[`An Item in review gets rendered well An Item in review with a labeled value gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <input
    id="select-form-item-item-key-1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="inline-label"
    htmlFor="select-form-item-item-key-1"
  >
    <ErrorTooltip
      additionalClassNames={
        Array [
          "small-8",
        ]
      }
    >
      <span
        className="small-8"
      >
        Item name
      </span>
    </ErrorTooltip>
    <span
      className="small-4 text-right"
    >
      Item value
      <small>
         (
        Item value label
        )
      </small>
      <a
        className="open-item-details margin-small-left"
        onClick={[Function]}
        title="Open item details"
      >
        <i
          className="fa fa-file-text-o"
        />
      </a>
    </span>
  </label>
</div>
`;

exports[`An Item in review gets rendered well An Item in review with a unit gets rendered well 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <input
    id="select-form-item-item-key-1"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="inline-label"
    htmlFor="select-form-item-item-key-1"
  >
    <ErrorTooltip
      additionalClassNames={
        Array [
          "small-8",
        ]
      }
    >
      <span
        className="small-8"
      >
        Item name
        <small>
           (
          kg
          )
        </small>
      </span>
    </ErrorTooltip>
    <span
      className="small-4 text-right"
    >
      Item value
      <a
        className="open-item-details margin-small-left"
        onClick={[Function]}
        title="Open item details"
      >
        <i
          className="fa fa-file-text-o"
        />
      </a>
    </span>
  </label>
</div>
`;

exports[`An Item in review gets rendered well An reviewed Item that was not submitted to edc 1`] = `
<div
  className="item padding-small-tb"
  id="id-1"
>
  <div
    className="disabled padding-small-rl"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <small
      style={
        Object {
          "display": "flex",
          "width": "100%",
        }
      }
    >
      <ErrorTooltip
        additionalClassNames={
          Array [
            "small-8",
          ]
        }
      >
        <span
          className="small-8"
        >
          Item name
        </span>
      </ErrorTooltip>
      <span
        className="small-4 text-right"
      >
        Item value
        <small>
           (
          Item value label
          )
        </small>
        <a
          className="open-item-details margin-small-left"
          onClick={[Function]}
          title="Open item details"
        >
          <i
            className="fa fa-file-text-o"
          />
        </a>
      </span>
    </small>
  </div>
</div>
`;
